src/matrix/room/timeline/persistence/GapWriter.js

```diff
@@ -47,28 +47,30 @@ export class GapWriter {
                 }
                 nonOverlappingEvents.push(...remainingEvents.slice(0, duplicateEventIndex));
                 if (!expectedOverlappingEventId || duplicateEventId === expectedOverlappingEventId) {
-                    // TODO: check here that the neighbourEvent is at the correct edge of it's fragment
-                    // get neighbour fragment to link it up later on
-                    const neighbourEvent = await txn.timelineEvents.getByEventId(this._roomId, duplicateEventId);
-                    if (neighbourEvent.fragmentId === fragmentEntry.fragmentId) {
-                        log.log("hit #160, prevent fragment linking to itself", log.level.Warn);
-                    } else {
+                    // Only link fragment if this is the first overlapping fragment we discover.
+                    // TODO is this sufficient? Might we get "out of order" fragments from events?
+                    if (!neighbourFragmentEntry) {
```

Comment:
I suspect it might not be sufficient ... in theory all the overlapping events can be distributed amongst many fragments.
We should probably take the most occuring fragment to link up with, and at the same time don't change any existing links? Perhaps also not connect to a fragment that we're already connected to (indrectly) in the other direction? This is complicated ... I need to think more about this.